README.md
# webdev

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-07-08
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-07-08
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/webdev/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/webdev/compare/0.1.0...0.1.1


src/webdev/item/view.clj
(ns webdev.item.view
  (:require [hiccup.page :refer [html5]] [hiccup.core :refer [html h]]))

(defn new-item []
  (html [:form.form-horizontal {:method "POST" :action "/items"}
         [:div.form-group.mb-2
          [:label.control-label.col-sm-2 {:for :name-input} "Name"]
          [:div.col-sm-10 [:input#name-input.form-control {:name :name :placeholder "Name"}]]]

         [:div.form-group.mb-2
          [:label.control-label.col-sm-2 {:for :desc-input} "Description"]
          [:div.col-sm-10 [:input#desc-input.form-control {:name :description :placeholder "Description"}]]]


         [:input.btn.btn-primary {:type :submit :value "New Item"}]]))

(defn delete-item-form [id]
  (html [:form {:method "POST" :action (str "/items/"id)}
         [:input {:type :hidden :name "_method" :value "DELETE"}]
         [:div.btn-group [:input.btn.btn-danger.btn-xs {:type :submit :value "Delete"}]]]))


(defn update-item-form [id checked]
  (html [:form {:method "POST" :action (str "/items/" id)}
         [:input {:type :hidden :name "_method" :value "PUT"}]
         [:input {:type :hidden :name "checked" :value (if checked "false" "true")}]
         [:div.btn-group [:button.btn.btn-primary.btn-xs (if checked "DONE" "TODO")]]]))

(defn items-page [items]
  (html5 {:lang :en}
         [:head [:title "Listronica"] [:meta {:name :viewport :content "width=device-width, initial-scale=1.0"}]
          [:link {:href "/bootstrap/css/bootstrap.min.css" :rel :stylesheet}]]
         [:body [:div.container
                 [:h1 "My Items"]
                 [:div.row (if (seq items)
                             [:table.table.table-striped [:thead [:tr [:th.col-sm-2][:th.col-sm-2][:th "Name"] [:th "Description"]]]
                              [:tbody (for [i items]
                                        [:tr [:td (delete-item-form(:id i))] [:td (update-item-form (:id i) (:checked i))] [:td (h (:name i))] [:td (h (:description i))]])]]

                             [:div.col-sm-offset-1 "There are no items."])]

                 [:div.col-sm-6 [:h2 "Create a new item"] (new-item)]]

          [:script {:src "https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"}]
          [:script {:src "/bootstrap/js/bootstrap.min.js"}]]))

src/webdev/item/model.clj
(ns webdev.item.model
  (:require [clojure.java.jdbc :as db]))

(defn create-table [db]
  (db/execute! db
               ["CREATE EXTENSION IF NOT EXISTS \"uuid-ossp\""])
  (db/execute! db
               ["CREATE TABLE IF NOT EXISTS items (
                 id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
                 name TEXT NOT NULL,
                 description TEXT NOT NULL,
                 checked BOOLEAN NOT NULL DEFAULT FALSE,
                 date_created TIMESTAMP NOT NULL DEFAULT now()
               )"]))

(defn create-item [db name desc]
  (:id (first (db/query db ["INSERT INTO items (name, description) VALUES (?, ?) RETURNING id" name desc]))))

(defn update-item [db id checked]
  (= [1] (db/execute! db ["UPDATE items SET checked=? where id=?" checked id])))

(defn delete-item [db id]
  (= [1] (db/execute! db ["DELETE FROM items where id=?" id])))

(defn read-items [db]
(db/query db ["SELECT id, name, description, checked, date_created FROM items ORDER BY date_created"]))

src/webdev/item/handler.clj
(ns webdev.item.handler
  (:require [webdev.item.model :refer [create-item delete-item read-items update-item]]
            [webdev.item.view :refer [items-page]]))

(defn handle-index-items[req]
  (let [db (:webdev/db req) items (read-items db)]
   {:status 200
    :body (items-page items)
    :headers {}}
    ))

(defn handle-create-item [req]
  (let [name (get-in req [:params "name"]) desc (get-in req [:params "description"]) db (:webdev/db req)
        item_id (create-item db name desc)]
    {:status 302
     :body ""
     :headers {"Location" "/items"}}))

(defn handle-delete-item [req]
  (let [db (:webdev/db req) item-id (java.util.UUID/fromString (:item-id (:route-params req)))
        exists? (delete-item db item-id)]
    (if exists? 
      {:status 302
       :body ""
       :headers {"Location" "/items"}}
      
       {:status 404
       :body "Item not found."
       :headers {}}
      )
    ))

(defn handle-update-item [req]
  (let [db (:webdev/db req) item-id (java.util.UUID/fromString (:item-id (:route-params req)))
        checked (get-in req [:params "checked"])
        exists? (update-item db item-id (= "true" checked))]
    (if exists?
      {:status 302
       :body ""
       :headers {"Location" "/items"}}

      {:status 404
       :body "Item not found."
       :headers {}})))

src/webdev/core.clj
(ns webdev.core

  (:require [ring.adapter.jetty :as jetty]
            [ring.middleware.reload :refer [wrap-reload]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.resource :refer [wrap-resource]]
            [compojure.core :refer [defroutes GET ANY POST PUT DELETE]]
            [compojure.route :refer [not-found]]
            [ring.handler.dump :refer [handle-dump]]
            [webdev.item.model :as items]
            [webdev.item.handler :refer [handle-index-items handle-create-item handle-delete-item handle-update-item]]))



(def db {:dbtype "postgresql"
         :dbname "webdev"
         :user "postgres"
         :password "123"})


(defn greet [req]
  {:status 200
   :body "hellooo"
   :headers {}})

(defn goodbye [req]
  {:status 200
   :body "bye"
   :headers {}})

(defn about [req]
  {:status 200
   :body "My name is Hareem Bilal"
   :headers {}})

(def ops {"+" +
          "-" -
          "*" *
          ":" /})

(defn calc [req]
  (let [a (Integer. (get-in req [:route-params :a]))
        b (Integer. (get-in req [:route-params :b]))
        op (get-in req [:route-params :op])
        f (get ops op)]
    (if f
      {:status 200
       :body (str (f a b))
       :headers {}}
      {:status 404
       :body (str "Unknown operator " op)
       :headers {}})))

(defn yo [req]
  (let [name (get-in req [:route-params :name])]
    {:status 200
     :body (str "Yo " name "!")
     :headers {}}))

(defroutes routes
  (GET "/" [] greet)
  (GET "/goodbye" [] goodbye)
  (GET "/about" [] about)
  (GET "/yo/:name" [] yo)
  (GET "/calc/:a/:op/:b" [] calc)
  (GET "/request" [] handle-dump)
  (GET "/items" [] handle-index-items)
  (POST "/items" [] handle-create-item)
  (DELETE "/items/:item-id" [] handle-delete-item)
  (PUT "/items/:item-id" [] handle-update-item)
  (not-found "page not found"))

(defn wrap-db [hdlr]
  (fn [req]
    (hdlr (assoc req :webdev/db db))))

(defn wrap-server [hdlr]
  (fn [req]
    (assoc-in (hdlr req) [:headers "Server"] "Listronica 9000")))

(def sim-methods {"PUT" :put
                  "DELETE" :delete})

(defn wrap-simulated-methods [hdlr]
  (fn [req]
    (if-let [method (and (= :post (:request-method req)) (sim-methods (get-in req [:params "_method"])))]
      (hdlr (assoc req :request-method method))
      (hdlr req)
      )))

(def app
  (wrap-server (wrap-resource (wrap-db (wrap-params (wrap-simulated-methods routes))) "static")))

(defn -main [port]
  (items/create-table db)
  (jetty/run-jetty app
                   {:port (Integer. port)}))

(defn -dev-main [port]
  (items/create-table db)
  (jetty/run-jetty (wrap-reload #'app)
                   {:port (Integer. port)}))